src/agent.rs

```diff
@@ -187,4 +210,130 @@ impl EmulatorCoreAgentState {
+                        val.to_string(),
+                    )))
+                    .await
+                    .unwrap();
+            }
+            Syscall::ReadInt => {
+                let scan_result = self.scanner.next_int();
+                match scan_result {
+                    None => {
+                        self.blocked_on = BlockedOn::Syscall(syscall);
+                    }
+                    Some(scan_result) => {
+                        self.blocked_on = BlockedOn::Nothing;
+                        match self.current_datapath.as_datapath_ref() {
+                            DatapathRef::MIPS(_) => {
+                                self.current_datapath.set_register_by_str("v1", scan_result);
+                            }
+                        }
+                    }
+                }
+            }
+            Syscall::ReadFloat => {
+                let scan_result = self.scanner.next_float();
+                match scan_result {
+                    None => {
+                        self.blocked_on = BlockedOn::Syscall(syscall);
+                    }
+                    Some(scan_result) => {
+                        self.blocked_on = BlockedOn::Nothing;
+                        match self.current_datapath.as_datapath_ref() {
+                            DatapathRef::MIPS(_) => {
+                                self.current_datapath
+                                    .set_fp_register_by_str("f0", f32::to_bits(scan_result) as u64);
+                            }
+                        }
+                    }
+                }
+            }
+            Syscall::ReadDouble => {
+                let scan_result = self.scanner.next_double();
+                match scan_result {
+                    None => {
+                        self.blocked_on = BlockedOn::Syscall(syscall);
+                    }
+                    Some(scan_result) => {
+                        self.blocked_on = BlockedOn::Nothing;
+                        match self.current_datapath.as_datapath_ref() {
+                            DatapathRef::MIPS(_) => {
+                                self.current_datapath
+                                    .set_fp_register_by_str("f0", f64::to_bits(scan_result));
+                            }
+                        }
+                    }
+                }
+            }
+            Syscall::ReadString(addr) => {
+                let scan_result = self.scanner.next_line();
+                match scan_result {
+                    None => {
+                        self.blocked_on = BlockedOn::Syscall(syscall);
+                    }
+                    Some(scan_result) => {
+                        self.blocked_on = BlockedOn::Nothing;
+
+                        let bytes = scan_result.as_bytes();
+                        let memory = self.current_datapath.get_memory_mut();
+                        for (i, byte) in (0..bytes.len()).zip(bytes) {
```

Comment:
We can also use `bytes.iter().enumerate()` here which may slightly reduce a bit of overhead since it uses one iterator instead of two. Unless there was a specific reason for using `(0..bytes.len()).zip(bytes)` ?